paddle/operators/detail/grpc_client.h

```diff
@@ -0,0 +1,138 @@
+#include <chrono>
+#include <ctime>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "paddle/framework/data_type.h"
+#include "paddle/framework/lod_tensor.h"
+#include "paddle/framework/scope.h"
+#include "paddle/framework/selected_rows.h"
+#include "paddle/operators/detail/simple_block_queue.h"
+#include "paddle/operators/detail/utils.h"
+
+namespace paddle {
+namespace operators {
+namespace detail {
+
+struct VarHandle {
+  std::string ep;
+  const platform::DeviceContext* ctx;
+  const framework::Scope* scope;
+  std::string name;
+
+  std::string String() const {
+    std::ostringstream s;
+    s << "name:[" << name << "] ep:[" << ep << "]";
+    return s.str();
+  }
+};
+
+void ProcSendResponse(const VarHandle&, const sendrecv::VoidMessage& msg);
+
+void ProcGetResponse(const VarHandle& var_h,
+                     const sendrecv::VariableMessage& msg);
+
+struct ClientBase {
+  explicit ClientBase(std::shared_ptr<grpc::Channel> ch) {
+    stub = sendrecv::SendRecvService::NewStub(ch);
+    context = NULL;
+  }
+
+  virtual ~ClientBase() {}
+
+  virtual void Prepare(const VarHandle& var_info, int64_t time_out) {
+    context.reset(new grpc::ClientContext());
+    var_h = var_info;
+
+    std::chrono::system_clock::time_point deadline =
+        std::chrono::system_clock::now() + std::chrono::milliseconds(time_out);
+
+    context->set_deadline(deadline);
+  }
+
+  virtual void Proceed() = 0;
+
+  std::unique_ptr<sendrecv::SendRecvService::Stub> stub;
+  std::unique_ptr<grpc::ClientContext> context;
+  grpc::Status status;
+  VarHandle var_h;
+};
+
+typedef std::function<void(const VarHandle&, const sendrecv::VoidMessage&)>
+    RequestSendCallBack;
+
+struct SendProcessor : public ClientBase {
+  explicit SendProcessor(std::shared_ptr<grpc::Channel> ch) : ClientBase(ch) {}
+
+  virtual ~SendProcessor() {}
+
+  virtual void Proceed() { response_call_back(var_h, reply); }
+
+  sendrecv::VoidMessage reply;
+  RequestSendCallBack response_call_back = ProcSendResponse;
+};
+
+typedef std::function<void(const VarHandle&, const sendrecv::VariableMessage&)>
+    RequestGetCallBack;
+
+struct GetProcessor : public ClientBase {
+  explicit GetProcessor(std::shared_ptr<grpc::Channel> ch) : ClientBase(ch) {}
+
+  virtual ~GetProcessor() {}
+
+  virtual void Proceed() { response_call_back(var_h, reply); }
+
+  sendrecv::VariableMessage reply;
+  RequestGetCallBack response_call_back = ProcGetResponse;
+};
+
+class RPCClient {
+ public:
+  bool AsyncSendVariable(const std::string& ep,
+                         const platform::DeviceContext& ctx,
+                         const framework::Scope* scope,
+                         const std::string& var_name,
+                         int64_t time_out = 600 * 1000);
+
+  bool AsyncGetVariable(const std::string& ep,
+                        const platform::DeviceContext& ctx,
+                        const framework::Scope* scope,
```

Comment:
The save as above
`const framework::Scope* scope `
=>
`const framework::Scope& scope `

Reply:
Done.